### PR TITLE
[all]: Add Load Signed Byte/Half Immediate/Register Instructions

### PR DESCRIPTION
**Pull request by lukeg101**
This patch adds support for the LDRSB/H AArch64 Instructions

### CODE REVIEW

**Comment by maranget** [comment]
Hi @lukeg101. Those two instructions perform sign extension from 8/16 bits to 32/64bits. For instance if location '[x]' of size 1 byte holds -1 (or 0xFF), then the target register of instruction LDRSB will hold -1 (or 0xFFFFFFFF).

The following test illustrates this point:
```
% cat LDRSB.litmus
AArch64 LDRSB
{ 
  0:X1 = x;
  int8_t x = -1;
}
    P0                  ;
 LDRSB W0,[X1] ;
forall (0:X0=-1)
```
I have checked by running the above test with litmus: teh final value of `0:X0` is -1 as an int (and 0xFFFFFFFF as an `int64_t). However, with the code of this PR we have:
```
% herd7  LDRSB.litmus
est LDRSB Required
States 1
0:X0=255;
No
...
```

The root cause is the use of the sign extension primitive with the target size (here `Word`, _i.e._ 32bits) where the source size would be appropriate (here `Byte`, _i.e._ 8bits). See comments in code for a suggested patch.

File: herd/tests/instructions/AArch64/A24.litmus [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+(* Tests ldrsb instruction *)
+{ 0:X0 = 1; 0:X1 = x; int8_t x = 127; }
+
+P0;
+  LDRSB X0, [X1];
+  MOV X2, #1;
+  ADD X0, X0, X2;
+
+exists (0:X0 = 129 /\ x = 0)
```
Comment (maranget) [inline herd/tests/instructions/AArch64/A24.litmus:11]: A successful final condition with a `forall` quantifier would be appreciated. For instance `forall 0:X0=128 /\ [x]=127`.

File: herd/tests/instructions/AArch64/A24.litmus [hunk 2]
```diff
@@ -0,0 +1,11 @@
+AArch64 A24
+
+(* Tests ldrsb instruction *)
+{ 0:X0 = 1; 0:X1 = x; int8_t x = 127; }
```
Comment (maranget) [inline herd/tests/instructions/AArch64/A24.litmus:4]: For this test to be used by litmus without a warning, it is better to declare `0:X0` to be used as a 64bits register: `int64_t 0:X0 =1`. Alternatively, one may not declare any type (implicit type is `int`, 32 bits) and use the`W` form or register un the code.

File: herd/tests/instructions/AArch64/A25.litmus [hunk 3]
```diff
@@ -0,0 +1,11 @@
+AArch64 A25
+
+(* Tests ldrsh instruction *)
+{ 0:X0 = 1; 0:X1 = x; int16_t x = 65535; }
```
Comment (maranget) [inline herd/tests/instructions/AArch64/A25.litmus:4]: Same suggestions as for test `A24`: either declare `0:X0` of type `int64_t` or use `W` registers in code.

**Review by maranget — CHANGES_REQUESTED** [review]
I suggest the following changes
1. Fix semantics (see suggested patch)
2. Add tests that illustrate the difference between signed (_.e.g._ `LDRSB`) and unsigned (_.e.g._ `LDRB`) loads. Suggestions follow.

**Comment by maranget** [comment]
Here are two tests that illustrate the difference between signed and unsigne loads:
```
AArch64 A26

(* 
{ 
  0:X1 = x;
  int8_t x = -1;
  int64_t 0:X4;
  int64_t 0:X6;
}

P0;
  LDRSB W0,[X1] ;
  MOV X4,X0     ;
  LDRB W2,[X1]  ;
  MOV X6,X2     ;
forall (0:X0=-1 /\ 0:X2=255 /\ 0:X4=4294967295 /\ 0:X6=255)
```
And
```
AArch64 A27

(* 
{ 
  0:X1 = x;
  int16_t x = -1;
  int64_t 0:X4;
  int64_t 0:X6;
}

P0;
  LDRSH W0,[X1] ;
  MOV X4,X0     ;
  LDRH W2,[X1]  ;
  MOV X6,X2     ;
forall (0:X0=-1 /\ 0:X2=65535 /\ 0:X4=4294967295 /\ 0:X6=65535)
```

**Comment by maranget** [comment]
Hi @lukeg101. Thanks for this PR. Fixed sized arithmetic and processor behaviours are often tricky and it is good to check  our implementation from time to time.

This useful PR has been neglected for too long and some rebasing is necessary. If too tricky, please tell.

**Comment by lukeg101** [comment]
Hi Luc, thank you for your feedback, I have addressed it, please let me know if there is something else :)

**Review by maranget — APPROVED** [review]

**Comment by maranget** [comment]
Thanks, merged.

**Pull request merged**